// conseiljs/dist/chain/tezos/contracts/TzbtcTokenHelper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TzbtcTokenHelper = void 0;
const jsonpath_plus_1 = require("jsonpath-plus");
const TezosConstants_1 = require("../../../types/tezos/TezosConstants");
const TezosTypes = __importStar(require("../../../types/tezos/TezosChainTypes"));
const TezosLanguageUtil_1 = require("../TezosLanguageUtil");
const TezosMessageUtil_1 = require("../TezosMessageUtil");
const TezosNodeReader_1 = require("../TezosNodeReader");
const TezosNodeWriter_1 = require("../TezosNodeWriter");
const TezosContractUtils_1 = require("./TezosContractUtils");
var TzbtcTokenHelper;
(function (TzbtcTokenHelper) {
    function verifyDestination(server, address) {
        return __awaiter(this, void 0, void 0, function* () {
            return TezosContractUtils_1.TezosContractUtils.verifyDestination(server, address, '187c967006ca95a648c770fdd76947ef');
        });
    }
    TzbtcTokenHelper.verifyDestination = verifyDestination;
    function verifyScript(script) {
        return TezosContractUtils_1.TezosContractUtils.verifyScript(script, 'ffcad1e376a6c8915780fe6676aceec6');
    }
    TzbtcTokenHelper.verifyScript = verifyScript;
    function getAccountBalance(server, mapid, account) {
        return __awaiter(this, void 0, void 0, function* () {
            const value = yield queryMap(server, mapid, `(Pair "ledger" 0x${TezosMessageUtil_1.TezosMessageUtils.writeAddress(account)})`);
            return Number((0, jsonpath_plus_1.JSONPath)({ path: '$.args[0].int', json: value })[0]);
        });
    }
    TzbtcTokenHelper.getAccountBalance = getAccountBalance;
    function getOperatorList(server, mapid) {
        return __awaiter(this, void 0, void 0, function* () {
            const value = yield queryMap(server, mapid, '"operators"');
            let addresses = [];
            for (const a of value) {
                addresses.push(TezosMessageUtil_1.TezosMessageUtils.readAddress(a.bytes));
            }
            return addresses;
        });
    }
    TzbtcTokenHelper.getOperatorList = getOperatorList;
    function getTokenMetadata(server, mapid = 31) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield queryMap(server, mapid, '"tokenMetadata"');
        });
    }
    TzbtcTokenHelper.getTokenMetadata = getTokenMetadata;
    function getSimpleStorage(server, address) {
        return __awaiter(this, void 0, void 0, function* () {
            const storageResult = yield TezosNodeReader_1.TezosNodeReader.getContractStorage(server, address);
            return {
                mapid: Number((0, jsonpath_plus_1.JSONPath)({ path: '$.args[0].int', json: storageResult })[0]),
                scale: 8
            };
        });
    }
    TzbtcTokenHelper.getSimpleStorage = getSimpleStorage;
    function getTokenSupply(server, mapid = 31) {
        return __awaiter(this, void 0, void 0, function* () {
            const r = yield queryMap(server, mapid, '"totalSupply"');
            return Number((0, jsonpath_plus_1.JSONPath)({ path: '$.int', json: r })[0]);
        });
    }
    TzbtcTokenHelper.getTokenSupply = getTokenSupply;
    function getPaused(server, mapid = 31) {
        return __awaiter(this, void 0, void 0, function* () {
            const r = yield queryMap(server, mapid, '"paused"');
            return ((0, jsonpath_plus_1.JSONPath)({ path: '$.prim', json: r })[0]).toLowerCase().startsWith('t');
        });
    }
    TzbtcTokenHelper.getPaused = getPaused;
    function transferBalance(server, signer, keystore, contract, fee, source, destination, amount, gas = 250000, freight = 1000) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = `(Pair "${source}" (Pair "${destination}" ${amount}))`;
            const nodeResult = yield TezosNodeWriter_1.TezosNodeWriter.sendContractInvocationOperation(server, signer, keystore, contract, 0, fee, freight, gas, 'transfer', parameters, TezosTypes.TezosParameterFormat.Michelson, TezosConstants_1.TezosConstants.HeadBranchOffset, true);
            return TezosContractUtils_1.TezosContractUtils.clearRPCOperationGroupHash(nodeResult.operationGroupID);
        });
    }
    TzbtcTokenHelper.transferBalance = transferBalance;
    function approveBalance(server, signer, keystore, contract, fee, destination, amount, gas = 250000, freight = 1000) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = `(Right (Right (Right (Right (Left (Right (Right (Right (Pair "${destination}" ${amount})))))))))`;
            const nodeResult = yield TezosNodeWriter_1.TezosNodeWriter.sendContractInvocationOperation(server, signer, keystore, contract, 0, fee, freight, gas, '', parameters, TezosTypes.TezosParameterFormat.Michelson, TezosConstants_1.TezosConstants.HeadBranchOffset, true);
            return TezosContractUtils_1.TezosContractUtils.clearRPCOperationGroupHash(nodeResult.operationGroupID);
        });
    }
    TzbtcTokenHelper.approveBalance = approveBalance;
    function mintBalance(server, signer, keystore, contract, fee, destination, amount, gas = 250000, freight = 1000) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = `(Right (Right (Right (Right (Right (Left (Left (Left (Pair "${destination}" ${amount})))))))))`;
            const nodeResult = yield TezosNodeWriter_1.TezosNodeWriter.sendContractInvocationOperation(server, signer, keystore, contract, 0, fee, freight, gas, '', parameters, TezosTypes.TezosParameterFormat.Michelson);
            return TezosContractUtils_1.TezosContractUtils.clearRPCOperationGroupHash(nodeResult.operationGroupID);
        });
    }
    TzbtcTokenHelper.mintBalance = mintBalance;
    function addOperator(server, signer, keystore, contract, fee, operator, gas = 250000, freight = 1000) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = `(Right (Right (Right (Right (Right (Left (Right (Left "${operator}" ))))))))`;
            const nodeResult = yield TezosNodeWriter_1.TezosNodeWriter.sendContractInvocationOperation(server, signer, keystore, contract, 0, fee, freight, gas, '', parameters, TezosTypes.TezosParameterFormat.Michelson);
            return TezosContractUtils_1.TezosContractUtils.clearRPCOperationGroupHash(nodeResult.operationGroupID);
        });
    }
    TzbtcTokenHelper.addOperator = addOperator;
    function queryMap(server, mapid, query) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = Buffer.from(TezosMessageUtil_1.TezosMessageUtils.writePackedData(query, '', TezosTypes.TezosParameterFormat.Michelson), 'hex');
            const packedKey = TezosMessageUtil_1.TezosMessageUtils.writePackedData(key, 'bytes');
            const encodedKey = TezosMessageUtil_1.TezosMessageUtils.encodeBigMapKey(Buffer.from(packedKey, 'hex'));
            const mapResult = yield TezosNodeReader_1.TezosNodeReader.getValueForBigMapKey(server, mapid, encodedKey);
            if (mapResult === undefined) {
                throw new Error(`Could not get data from map ${mapid} for '${query}'`);
            }
            const bytes = (0, jsonpath_plus_1.JSONPath)({ path: '$.bytes', json: mapResult })[0];
            return JSON.parse(TezosLanguageUtil_1.TezosLanguageUtil.hexToMicheline(bytes.slice(2)).code);
        });
    }
})(TzbtcTokenHelper = exports.TzbtcTokenHelper || (exports.TzbtcTokenHelper = {}));
